### ISSUE TITLE
Allow an application to add iteration spaces

### ISSUE THREAD

**Issue by hiker** [stfc/PSyclone#191]
If an application needs a new iteration space, at this stage a modification of psyclone is required, which is rather inconvenient.

I have started to add the ability for an application to add iteration spaces for GOcean1.0. Within psyclone, there is a simple function call, e.g.:
```
    GOLoop.add_bounds("offset_sw:cu:north-halo:{start}:{stop}:1:1")
```

So it's offset:field-type:iteration-space:inner-start:inner-stop:outer-start:outer-stop (maybe order of outer and inner loop boundaries should be changed ... not sure).

Now in order for an application to call this function and add new iteration spaces for its use in PSyclone, I intend to add an additional command line parameter, e.g. ```--config-file  some/where/config```. The config file would contain something like:
```
...
[iteration-space]
   offset_sw:cu:north-halo = {start}:{stop}:1:1
   offset_sw:cu:south-halo = {start}:{stop}:{stop}:{stop}
   ...
```
PSyclone would  then parse this config file and add all specified iteration spaces to GOcean1p0. Then the application can use these new iteration spaces as expected.

Considering that there already is a config file, would it be worth combining those two? Unfortunately that's not entirely easy - e.g. we would need to support a command line option to specify a config file to be loaded, and atm before optparse is called the config files already needs to be read (in order to define supported APIs) :) So from this point of view it appears to be quite different config files, but I thought I better check with you. 

Secondly, I like the ":" as separator, but by default the config parser interprets the first ':' as '=', so e.g.:
```
a:b:c = 1:2:3:4
a:b:d = 5:6:7:8
```
is actually not valid - it will assign 'b:c = 1:2:3:4" to the key 'a', and then find another value for 'a' (b:c:d = 5:6:7:8). I could either change this (the configfile constructors allows me to disable the ":"), or use a different symbol as delimiter ... I just can't find a good one: '-' might work, but we can't use '-' in the loop boundaries (since we might have a loop to {stop}-1 or so). The symbol '_' somehow feels wrong (not to mention that it might interfere with a Fortran type value, 1_single or so); '@' looks odd ... any preferences? I would tend to just disable ":" as delimiter for the config file.

Any other things you want me to look out for/support?

**Comment by arporter** [comment]
This sounds good. I am quite happy to disable ":" as a delimiter - we just need to document the fact that we've done it in the relevant section of the user guide.

I would prefer to add the iteration-space definitions in the existing config file if possible, maybe under an API-specific section. It may be possible to make `psyclone` take the location of a script from the command line - with luck it is just a matter of moving the location of the line that creates the initial config object in `generate.py`. However, as I say, I think it would be best if we had just the one config file kept either in the standard locations or pointed to by PSYCLONE_CONFIG.

While I'm here, note that 'cu' is not actually the field type but specifies that the field values are located at the U points on an Arakawa C grid (but then, you probably already knew that...).

**Comment by hiker** [comment]
I had a long think about the config file. My 'problem' is that the current config file is a mixture between a 'package' file (containing information about a package), and a typical user config file:

1. It contains the list of valid APIs, which should not be edited by the user - the list of valid APIs is part of the package.
2. The other settings like defaults etc are more user specific settings.
3. The file is looked for at different locations (which would indicate this to be more a user config file) - but it does not look e.g. in .. or ../.. (e.g. the way a .git directory would be found) or $HOME/.psyclone (or $XDG_CONFIG_HOME). When I got the message about missing .psyclone/psyclone.cfg, my first impulse was to create $HOME/.psyclone/psyclone.cfg ... which is imho kind of standard, but does not work.

Also, the fact that the config file is read before defining the parser of the command line (in order to get the list of valid APIs) makes it hard to define a new config file on the command line (as far as I can see we would need to handle some part of the validation that is being done in argparse manually).

TBH, I think thinking about this should be a separate issue (*grin* and then we can also discuss using a different file format ... this format is a pain, see below ;)  ) - so, I suggest to do a simpler solution for this ticket here:

I add support for another config file (based on the existing file format) using a --config command line option. This file is read after command line parsing is done (since it is specified as a command line option). It contains a API specific section ('[gocean1.0]'), and will be read in and the boundary specification will be added to the GOcean api. Very simple and clean to implement.

If we should decide to have multiple config files (e.g. a package specific file, followed by user-specific files from $HOME/.psyclone, $XDG_CONFIG_HOME, .., ../.. ...) the reading code can be moved into the the current config file reader (well, creating a new GOceanConfig class). Note that the python config parser already supports reading more than one file - we can either specify a list of files, or call read() more than once - and any later read file will overwrite any settings from a previous file.

Or if we really want only one config file, again we can just move the section and reading as above.

Since the file format is rather limited: we need to use a [gocean1.0] section, we can't nest sections e.g. to have [iteration-spaces] under [gocean1.0]), and each keyword can only be specified once. This leaves as options:
1. Use ```iteration-space-1=...``` , ```iteration-space-2=...``` etc - which I somehow find ugly
2. Use a multi-line specification:
```
iteration-space = spec1
    spec2
    spec3
```
Unless someone comes up with a better solution, I'll implement the latter.

And an iteration-space specification would for example:
```
iteration-space=offset_ne:ct:all_pts:1:{stop}+1:1:{stop}+1
```
So it's offset-type:grid-type:iteration-space:outer-start:outer-stop:inner-start:inner-stop

Or I could of course a less dense but more descriptive way, e.g.:
```
iteration-space=offset=offset_ne grid-type=ct iterationspace=all_pts outer_start=1...
```
In any way I need to add the parsing and error handling for whatever format we use :( I'll use the first syntax, again unless you prefer something else.

It would be so much nicer with xml - but let's not go there ;)

**Comment by arporter** [comment]
There's a lot to think about there! I take your point about the ~dual purpose nature of the current config file. I think we could simplify some of the problems by simply taking the list of supported APIs out of that file and putting them into the PSyclone source - as you say, they are not a user-configurable option.
With that change we could probably read the config file *after* the command-line options have been parsed.

The reason the config file isn't searched for in ~/.psyclone/ (which was also my first choice) is to do with difficulties handling virtual environments. Basically, robustly handling the various ways a package can be installed is a real pain :-( Having said all that, I'm not sure why I didn't leave ~/.psyclone as an option. Maybe something to revisit.)

Multi-line looks like the best option available to us without dropping configparser. I'll have a think about single versus multiple config files over lunch and get back to you.

**Comment by arporter** [comment]
I've done some more thinking and have decided I would prefer to keep a single config file. Reading back through this Issue, I can't quite understand why you wanted to add an additional one - what was the reasoning?

**Comment by hiker** [comment]
Because of:
```
    parser.add_argument(
        '-api', default=_CONFIG.default_api,
        help='choose a particular api from {0}, '
        'default {1}'.format(str(_CONFIG.supported_apis), _CONFIG.default_api))
```
This requires that the config file was read before any command line is handled.

So in order to have only one config file: I'll move the supported APIs out of the config file (but leave it in the config object - imho that makes sense). This then still leaves the problem of defining a default API - which now should come from the potentially specified config file. So I'll remove _CONFIG.default_api as well from the optparser arguments, read one config file (either the one specified on the command line, or otherwise the one that is currently found), then detect if no API is set I'll pick the default from the config file that was just read. The only problem I see with this is that psyclone --help can not display the default API anymore (since it now depends on the config file, which has not been read at the time the arguments are defined).

Or am I missing something? Maybe we can detect if --help was specified, and if so, manually print the default API, since now we can read a config file??

**Comment by arporter** [comment]
I don't think detecting whether `--help` was specified is going to be possible without messing with the internals of the argument parser. I'm happy to drop specifying what the default API is in the help message if that's the price we pay.

@rupertford has suggested that, since we're going down this approach, we could (in a subsequent issue) move all of the iteration space configuration out of PSyclone and into the config file. i.e. the `internal_pts` and `all_pts` iteration spaces for each offset and field type. I'm slightly concerned that I can't remember how this relates to the implementation in the `dl_esm_inf` library. We should therefore write some developers' documentation that describes how this definition of an iteration space is used by PSyclone when generating the PSy layer!

**Comment by hiker** [comment]
OK, I've done some 'keyhole surgery': I moved the supported APIs (and stubs) into the Config constructor (and do not read them from the config file anymore). So we can use the config object to still print the allowed APIs in the help message for argparse (which I found quite useful when I started). Then after command line handling is done, load the actual config file (either the user specified one, or the default). I use a special 'do not read config file' parameter when creating the config instance (and later call a Config.load(...) function). This way all tests keep on working as expected (since all other uses of Config except in generator are unchanged :)

Unfortunately I found two problems:
1) The default for '-api' is also taken from the config object. Since the config object has not been read, we can't do that anymore. Potential solution: 
    - Don't set a default for API. Easy to do, and imho makes sense (long term each project will just have a local config file, which specifies the API to use, so no need for the API parameter at all).
    - We first 'manually' check the command line options for a '--config', load the config file and remove the option from the command line parameters. Then normal optparse handling will work fine.

I don't have much of a preference for either tbh.
2) Much much bigger problem: The line
```from psyclone.psyGen import PSyFactory, GenerationError```
in generator.py will actually create a config instance (since _CONFIG=... is in the global name space of psyGen - it is executed even though we don't import _CONFIG :( ). So even though I suppress reading of the config file in generator.py, it is actually still read when doing the above import line since _CONFIG is global :( Basically, my refactoring actually does not work at this stage :(

This is unfortunately more of an issue (since it means when defining the command line parameter, a Config file has already been read).  Potential solutions:
1) We move all those imports in generator.py away from the top of the file into the functions where they are actually used. This way no config object is created implicitly since the global import from psyGen (and others) do not happen anymore, and my change would work as expected. Disadvantage: we would have potentially several identical imports in different functions (since we might not know the order in which functions are called, so we need to import from psyGen everywhere - e.g. GenerationError from psyGen is used in a lot of places), this is messy to maintain, not to mention I am not 100% sure if this is actually possible (we might need some objects from an import somewhere before we have read the config file).
2) We disable reading of a config file from all those global _CONFIG=ConfigFactory().create() lines everywhere. Disadvantage: I would expect problems with tests, since they likely need a loaded Config file, which means we would need to add explicit loading statements in many (most?) tests :( Not good.
3) We get rid of the global _CONFIG variables.

The last option seems to me the best solution (though suggestions are welcome ;) ). Which raises another question:
it would be easy enough to replace all global _CONFIG variables with local ones (i.e. creating them in all functions using the config factory) ... but this shows imho some disadvantages of the current implementation, since we then will have to create many more instances of the Factory, just to get access to the instance - i.e. more overhead in terms of time and memory handling. My suggestion: we remove the Factory, since I think it's a bit odd to use a factory just to get a singleton instance. Could we just use:
```
Config.get()->distributed_memory 
```
The get method would return one instance (and create one if no instance is defined), but avoid the creation of a ConfigFactory instance in between, making it a lot easier to get the instance (and we would still be able to store a local copy if we have many calls to a config method in a function: _config = Config.get()).

Hope that explanation makes sense, I have been staring at this for a while now :)

**Comment by hiker** [comment]
> @rupertford has suggested that, since we're going down this approach, we could (in a subsequent issue) move all of the iteration space configuration out of PSyclone and into the config file

Ah, Rupert has spotted my secret master plan ... 50% of it :) Yes, that part is obviously the next step ... but I actually wanted to take it a step further and potentially move the whole handling of offsets out of PSyclone :) So instead of two 'attributes' (one for the field type like ct, cf, ... and one for the offset), we would only have one field type, and in case of dl_esm_inf if could be a combined key: ct_se etc. This should make not much difference to existing code (well ... some, but not too bad), since we can't mix offset types anyway, and each application can then define exactly the data (offset + field types) it needs, and nothing more.

The advantage is that this would make it a lot easier to use PSyclone and dl_esm_inf with a code that does not need the offset type in the first place :) 

But ... this master plan is secret, so don't tell anyone (ok, to be serious: I haven't even start looking if this is de-facto feasible. This enhancement is useful on its own, and I will check this later ).

**Comment by hiker** [comment]
@arporter, @rupertford I need some feedback here, since I am kind of stuck:
The global variables are really a pain^h^h^h^h^h^h annoying :)
I have most things properly implemented and tested, the gocean config class is available as an API from the configuration class, and as far as I can see if works fine (including the tests for the new code, and from the actual command line) ... but it does not play nice with the existing test cases (and only those as far as I can tell).

Issue is that some of the config tests load new config files. So initially I added a teardown function to the config tests which would just delete the singleton, and rely on the various classes to reload the default one. But some classes store the Config issue as a global variable - and even though I free the config instance, those global pointers in various files keep the old instance alive. E.g.
1) an import psyGen creates a global _CONFIG variable in psyGen. 
2) Then a test loads a new one config file (second instance).
3) At the end of the tests I delete this instance in the test files and load a the default config file again (2nd one freed, 3rd created ... and 1st sill exists).

So I ended up with two instances of the singleton, one in (e.g) psyGen, one in the test functions. Some tests modify the config file directly (e.g. for the annexed dof tests), and what happens then is that different files have different instances of the config _singleton_ - some created when a module was imported (even though not even used), others later. So this caused quite a few test failures, and quite a bit of a mess (took me a while to figure out what is happening ... two instances of singletons *sigh*). Those test failures depend on the order in which the tests are executed - quite nasty. This issue I believe already exist in master (I only have a vague idea why your unit tests are still working ... sheer coincidence of the order in which modules are imported ... I think):
```
py.test   ./dynamo0p3_transformations_test.py 
,,,
collected 165 items 

dynamo0p3_transformations_test.py .............
```
but when config_test is executed first:
```
py.test ./config_test.py  ./dynamo0p3_transformations_test.py 
...
collected 182 items 

config_test.py .................
dynamo0p3_transformations_test.py FFFFF........
...
E           TransformationError: 'Transformation Error: Error in DynamoColour transformation. The supplied node is not a loop'

```
Reason is (from memory, I have debugged too many failures ;)  ) that the test case set annexed_dofs, but somewhere in dynamo0p3 or transform or so it does use a different instance of the dynamo config class, which has a wrong value.

So instead I decided to just clean/reload the config instance . This way all modules have pointers to the same instance, and I though that would solve the issue, but no: it turns out that e.g. dynamo0p3_transformations_test does:
```
_CONFIG = ConfigFactory().create()
_API_CONFIG = _CONFIG.api(TEST_API)
```
as global variables. So _API_CONFIG is initialised whenever that module is imported first. But when the config tests are executed, and do the cleanup (reload the default file) at the end, a new API instance is created, so again there are two instances - this time not of the Config class, but the API specific Config instance - one in the test class, and a different one in the psyclone classes.

I can't see an easy solution :( As far as I can tell removing all those global variables would work, which also means (imho) to change the singleton to get rid of the factory (to reduce number of calls):
```
class Config():
   _instance = None
   @staticmethod
   def get():
       if Config._instance is None:
            Config._instance = Config()
       return Config._instance
```
(etc etc ... just go get an idea) which means we could simply use (1) instead of (2):
```
Config.get()->distributed_memory,    # (1)
ConfigFactory().create().distributed_memory  # (2)
```
Suggestions welcome.

**Comment by arporter** [comment]
I remember being wary of the global configuration object when I introduced it but, as it didn't introduce any major difficulties, I kept it. I think I did work around difficulties in some tests by manually creating a Config object myself rather than using the Factory - this meant that the 'singleton' instance wasn't used.
I'm happy to get rid of the factory (I confess I copied Rupert's namespace manager which, I suspect, he copied from somewhere else ;-) ). Your solution is nicer. I'm also happy to get rid of the global instances of _CONFIG (and _API_CONFIG) as required.
